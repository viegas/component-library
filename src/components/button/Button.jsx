import React from 'react';
import PropTypes from 'prop-types';

import * as Styled from './Button.styles';
import Loader from '../loader/Loader';

const Button = ({ type, disabled, busy, children, ...props }) => {
    const blockActions = disabled || busy;

    return (
        <Styled.Button {...props} type={blockActions ? 'disabled' : type}>
            {children}
            {busy && (
                <Styled.LoaderWrapper>
                    <Loader color="blue-700" />
                </Styled.LoaderWrapper>
            )}
        </Styled.Button>
    );
};

Button.propTypes = {
    type: PropTypes.oneOf(['primary', 'secondary']),
    disabled: PropTypes.bool,
    busy: PropTypes.bool,
    children: PropTypes.node,
};

Button.defaultProps = {
    type: 'primary',
    disabled: false,
    busy: false,
    children: null,
};

export default Button;
