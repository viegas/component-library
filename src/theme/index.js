const theme = {
    spacings: {
        small: '5px',
        medium: '10px',
        large: '20px',
    },
    font: {
        size: {
            info: '12px',
            text: '16px',
            subTitle: '18px',
            title: '24px',
        },
        weight: {
            ligth: 300,
            regular: 400,
            semibold: 600,
            bold: 700,
            extraBold: 800,
        },
    },
    colors: {
        white: '#fff',
        black: '#0c0d0e',
        orange: '#f48024',
        blue: '#0077cc',
        'blue-100': '#cfeafe',
        'blue-200': '#addafc',
        'blue-300': '#6cbbf7',
        'blue-400': '#379fef',
        'blue-500': '#0095ff',
        'blue-600': '#0077cc',
        'blue-700': '#0064bd',
        'blue-800': '#0054a3',
        'blue-900': '#004487',
        powder: '#e1ecf4',
        'powder-050': '#f4f8fb',
        'powder-100': '#e1ecf4',
        'powder-200': '#d1e5f1',
        'powder-300': '#b3d3ea',
        'powder-400': '#a0c7e4',
        'powder-500': '#7aa7c7',
        'powder-600': '#5b8db1',
        'powder-700': '#39739d',
        'powder-800': '#2c5777',
        'powder-900': '#1e3c52',
        green: '#5eba7d',
        'green-025': '#eef8f1',
        'green-050': '#dcf0e2',
        'green-100': '#cae8d4',
        'green-200': '#a6d9b7',
        'green-300': '#82ca9a',
        'green-400': '#5eba7d',
        'green-500': '#48a868',
        'green-600': '#3d8f58',
        'green-700': '#2f6f44',
        'green-800': '#29603b',
        'green-900': '#1e472c',
        yellow: '#fbf2d4',
        'yellow-050': '#fdf7e3',
        'yellow-100': '#fbf2d4',
        'yellow-200': '#f1e5bc',
        'yellow-300': '#e6d178',
        'yellow-400': '#e9c63f',
        'yellow-500': '#ddb624',
        'yellow-600': '#cea51b',
        'yellow-700': '#b89516',
        'yellow-800': '#9f8010',
        'yellow-900': '#826a0b',
        red: '#d1383d',
        'red-050': '#fdf3f4',
        'red-100': '#f9d3d7',
        'red-200': '#f4b4bb',
        'red-300': '#e87c87',
        'red-400': '#de535e',
        'red-500': '#d1383d',
        'red-600': '#c02d2e',
        'red-700': '#ac2726',
        'red-800': '#942121',
        'red-900': '#7a1819',
        gold: '#ffcc01',
        'gold-lighter': '#fff4d1',
        'gold-darker': '#f1b600',
        silver: '#b4b8bc',
        'silver-lighter': '#e8e8e8',
        'silver-darker': '#9a9c9f',
        bronze: '#caa789',
        'bronze-lighter': '#f2e9e1',
        'bronze-darker': '#ab825f',
        transparent: '#ffffff00',
    },
    media: {
        breakpoints: {
            md: 768,
            lg: 1024,
        },
    },
};

export const mediaQueriesGenerator = (breakpoints) =>
    Object.keys(breakpoints).reduce((acc, breakpoint) => {
        const size = theme.media.breakpoints[breakpoint];
        const mediaQuery = `(min-width: ${size}px)`;

        acc[breakpoint] = `@media ${mediaQuery}`;
        return acc;
    }, {});

theme.media.query = mediaQueriesGenerator(theme.media.breakpoints);

export default theme;
