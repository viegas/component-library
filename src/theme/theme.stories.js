import React from 'react';
import { storiesOf } from '@storybook/react';

import theme from './index';

const boxStyle = {
    padding: '20px',
    paddingTop: '0px',
    marginBottom: '20px',
    borderBottom: '1px solid gray',
};

const Theme = () => {
    return (
        <div>
            <div style={boxStyle}>
                <p>Spacings</p>
                <div style={{ backgroundColor: 'gray', overflow: 'hidden' }}>
                    {Object.keys(theme.spacings).map((key) => (
                        <div
                            style={{
                                backgroundColor: 'white',
                                marginLeft: theme.spacings[key],
                                marginBottom: theme.spacings[key]
                            }}
                        >
                            {key} - {theme.spacings[key]}
                        </div>
                    ))}
                </div>
            </div>
            <div style={boxStyle}>
                <p>Colors</p>
                <table>
                    <tbody>
                        {Object.keys(theme.colors).map((key) => (
                            <tr>
                                <td
                                    style={{
                                        background: theme.colors[key],
                                        padding: '20px',
                                        width: '33%',
                                    }}
                                />
                                <td
                                    style={{
                                        color: theme.colors[key],
                                        padding: '20px',
                                    }}
                                >
                                    Aa
                                </td>
                                <td
                                    style={{
                                        padding: '20px',
                                    }}
                                >
                                    {key} - {theme.colors[key]}
                                </td>
                            </tr>
                        ))}
                    </tbody>
                </table>
            </div>
            <div style={boxStyle}>
                <p>Font Size</p>
                {Object.keys(theme.font.size).map((key) => (
                    <div style={{ fontSize: theme.font.size[key] }}>
                        {key} - {theme.font.size[key]}
                    </div>
                ))}
            </div>
            <div style={boxStyle}>
                <p>Font Weights</p>
                {Object.keys(theme.font.weight).map((key) => (
                    <div style={{ fontWeight: theme.font.weight[key] }}>
                        {key} - {theme.font.weight[key]}
                    </div>
                ))}
            </div>
        </div>
    );
};

storiesOf('Styles', module).add('theme.js', () => {
    return <Theme />;
});
