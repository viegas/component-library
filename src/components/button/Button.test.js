import Button from './Button';
import * as Styled from './Button.styles';

import theme from '../../theme';

describe('<Button />', () => {
    it('Should render correctly', () => {
        const wrapper = shallow(<Button>Test</Button>);

        expect(wrapper).toMatchSnapshot();
    });
    it('Should render correctly with disabled prop', () => {
        const wrapper = shallow(<Button disabled>Test</Button>);

        expect(wrapper).toMatchSnapshot();
    });

    it('Should render correctly the busy loader', () => {
        const wrapper = shallow(<Button busy>Test</Button>);

        expect(wrapper).toMatchSnapshot();
    });

    describe('<Styled.Button />', () => {
        it('Check primary type', () => {
            const wrapper = renderStyledComponent(<Styled.Button type="primary" />);

            expect(wrapper).toHaveStyleRule('border', `solid 2px ${theme.colors.blue}`);
            expect(wrapper).toHaveStyleRule('background-color', theme.colors.blue);
            expect(wrapper).toHaveStyleRule('color', theme.colors.white);

            expect(wrapper).toMatchSnapshot();
        });

        it('Check secondary type', () => {
            const wrapper = renderStyledComponent(<Styled.Button type="secondary" />);

            expect(wrapper).toHaveStyleRule('border', `solid 2px ${theme.colors.blue}`);
            expect(wrapper).toHaveStyleRule('background-color', theme.colors.white);
            expect(wrapper).toHaveStyleRule('color', theme.colors.blue);

            expect(wrapper).toMatchSnapshot();
        });

        it('Check disabled type', () => {
            const wrapper = renderStyledComponent(<Styled.Button type="disabled" />);

            expect(wrapper).toHaveStyleRule('border', `solid 2px ${theme.colors['silver-darker']}`);
            expect(wrapper).toHaveStyleRule('background-color', theme.colors['silver-lighter']);
            expect(wrapper).toHaveStyleRule('cursor', 'not-allowed');

            expect(wrapper).toMatchSnapshot();
        });
    });

    styledComponentsSnapshots(Styled, ['LoaderWrapper']);
});
