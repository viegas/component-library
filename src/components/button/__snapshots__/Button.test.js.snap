// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> <Styled.Button /> Check disabled type 1`] = `
.c0 {
  position: relative;
  max-width: 320px;
  border-radius: 50vh;
  padding: 5px 20px;
  font-size: 12px;
  text-align: center;
  text-transform: uppercase;
  white-space: nowrap;
  overflow: hidden;
  cursor: pointer;
  border: solid 2px #9a9c9f;
  background-color: #e8e8e8;
  cursor: not-allowed;
}

<button
  className="c0"
  type="disabled"
/>
`;

exports[`<Button /> <Styled.Button /> Check primary type 1`] = `
.c0 {
  position: relative;
  max-width: 320px;
  border-radius: 50vh;
  padding: 5px 20px;
  font-size: 12px;
  text-align: center;
  text-transform: uppercase;
  white-space: nowrap;
  overflow: hidden;
  cursor: pointer;
  border: solid 2px #0077cc;
  background-color: #0077cc;
  color: #fff;
}

.c0:hover {
  background-color: #0095ff;
}

<button
  className="c0"
  type="primary"
/>
`;

exports[`<Button /> <Styled.Button /> Check secondary type 1`] = `
.c0 {
  position: relative;
  max-width: 320px;
  border-radius: 50vh;
  padding: 5px 20px;
  font-size: 12px;
  text-align: center;
  text-transform: uppercase;
  white-space: nowrap;
  overflow: hidden;
  cursor: pointer;
  border: solid 2px #0077cc;
  background-color: #fff;
  color: #0077cc;
}

.c0:hover {
  border-color: #379fef;
  color: #379fef;
}

<button
  className="c0"
  type="secondary"
/>
`;

exports[`<Button /> <Styled.LoaderWrapper /> Should render correctly 1`] = `
.c0 {
  position: absolute;
  left: 50%;
  top: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c0 > svg {
  display: block;
}

<span
  className="c0"
/>
`;

exports[`<Button /> Should render correctly 1`] = `
<Buttonstyles__Button
  type="primary"
>
  Test
</Buttonstyles__Button>
`;

exports[`<Button /> Should render correctly the busy loader 1`] = `
<Buttonstyles__Button
  type="disabled"
>
  Test
  <Buttonstyles__LoaderWrapper>
    <Loader
      color="blue-700"
      size="sm"
    />
  </Buttonstyles__LoaderWrapper>
</Buttonstyles__Button>
`;

exports[`<Button /> Should render correctly with disabled prop 1`] = `
<Buttonstyles__Button
  type="disabled"
>
  Test
</Buttonstyles__Button>
`;
